### DISCUSSION TITLE
Options for dealing with desync

### DISCUSSION THREAD

**Discussion by TS-x** [mbround18/valheim-docker#257] [General]
Like many others, my server too is experiencing massive desync with chests, pickups, NPCs etc. with 6 or more players connected. As I understand it, this is largely to do with how the game offloads calculations to clients and latency problems. Obviously this is a development issue unrelated to docker, but I'm curious to know what specific measures people using this image have taken to try to mitigate it.

I installed BepInEx and this(https://www.nexusmods.com/valheim/mods/55) modification, but had no success. I'm unsure whether the mod itself is ineffectual, or if the next step is to try and get all clients to use modified DLLs and hope it works in tandem with the server side fix. I'd expect that might drive some of them away from playing, and complicate things with future updates.

Right now all we can do is try our best to avoid gathering in larger groups and constructing large encampments, but that sucks a lot of fun out of the game. Any advice would be appreciated. 

Thanks!

**Comment by GSchuringa** [comment]
What kind of device are you running the server on?

**Reply by TS-x** [reply]
A home server running Debian/3.8GHz quad core/8GB RAM/SSD and 30Mbit upload. It's a little dated now, so I hope I wasn't too eager in thinking I could self-host. During our peak hours, I see minimal CPU usage and 2-3GB of RAM allocated to Valheim, with an additional 3GB still free.

**Reply by TheDakk** [reply]
I'm pretty sure the issue is with the netcode in the game itself, you can read more about it here:

https://valheimbugs.featureupvote.com/

There's no other fix that I've seen besides the DLL hack.

**Reply by spannerman79** [reply]
> A home server running Debian/3.8GHz quad core/8GB RAM/SSD and 30Mbit upload. It's a little dated now, so I hope I wasn't too eager in thinking I could self-host. During our peak hours, I see minimal CPU usage and 2-3GB of RAM allocated to Valheim, with an additional 3GB still free.

I don't think its your specs - as long as your load average doesn't exceed your total threads/cores (ie 4c/2t = 8 total threads, cannot exceed load avg 8.0).

Valheim is still based too much on P2P networking.  Even though they have "imporved" it ping from everyone still matters.

And it will only get worse if they do decide to switch to an authoritative server model.  That is ideal in order to stay away from any sort of anti cheat.  But authoritative server models are heavily ping dependant.

Non authoritative server models are those that need anti cheat. And as a linux user anti cheats are a bane.

**Comment by wrparker** [comment]
Hey @TS-x , I was having similar issues to you running this on a dedicated VPS.  I also had issues with that same BepInex mod working.  The desync was worse than if someone just hosted it locally.    I think I might've finally found a solution that directly updates the `m_dataPerSec` value in the dll without BepInEx mods.

**Here is my source (Reddit Post)**
https://www.reddit.com/r/valheim/comments/lr4qh0/how_to_fix_assembly_valheimdll_in_linux_based/

**Script**
https://github.com/nchie/valheim-desync-fix/blob/main/apply-valheim-lagfix.py

I can write something up on how to apply this to this to the installation files on the docker image (i forget where the mount point is).  I imagine every-time the game updates you'll have to re-patch (or any time the update replaces the assembly_valheim.dll file).

**EDIT**: To be clear we were able to achieve performance improvements only applying this patch to the server without any client modification.

**Reply by spannerman79** [reply]
To be fair ValheimPlus does do this correctly ( via BepInEx - since 0.8.5)

https://github.com/valheimPlus/ValheimPlus/blob/development/ValheimPlus/GameClasses/ZDOMan.cs#L19

Defined (in config) via https://github.com/valheimPlus/ValheimPlus/blob/development/valheim_plus.cfg#L436..L437

**Reply by wrparker** [reply]
Thanks for your reply!  I never made an attempt to do this with ValheimPlus, but I'm sure it handles this better than the mod TS-x and I have both tried. 

Unfortunately, My users aren't modders (non-technical) and from what I could tell VP+ requires installation on sever & clients. I am shooting for easiest fix possible without requiring client side mods. So far, I seem to be having success manually patching my dll.  I'll also add that we've only been able to test desync/lag relief with 4-5 concurrent users. I can't speak for if this works past ~5 users or if you're better off using modded server + clients. 

If my users were modders though and not completely intimidated by it, I'd give VP+ a shot because I'm sure there are other QoL improvements that come with it.

**Reply by spannerman79** [reply]
Its pretty easy to mod.

https://valheim.thunderstore.io/

And the mod manager is cross platform too https://github.com/ebkr/r2modmanPlus/releases/latest - unlike that Vortex cancer.

As to when V+ hits Thunderstore - https://github.com/valheimPlus/ValheimPlus/issues/138#issuecomment-786254905

>If my users were modders though and not completely intimidated by it, I'd give VP+ a shot because I'm sure there are other QoL improvements that come with it.

Oh more then what you know.  Auto Depost for processed kiln, furnace etc so it doesn't just spew out into the world.  Map sharing (only when others are online at the same time), plus a bunch more with just a single mod.

**Reply by TS-x** [reply]
> To be fair ValheimPlus does do this correctly ( via BepInEx - since 0.8.5)
> 
> https://github.com/valheimPlus/ValheimPlus/blob/development/ValheimPlus/GameClasses/ZDOMan.cs#L19
> 
> Defined (in config) via https://github.com/valheimPlus/ValheimPlus/blob/development/valheim_plus.cfg#L436..L437

I've experimented with V+ running on a different image with native support. While I can definitely see an increased send rate (sometimes exceeding 350kb) from the server via F2, it still hasn't helped to reduce the desync, even with as few as 3 players in a zone, and it creates an additional problem of its own. With the data rate configuration set anywhere from 128-500, it begins to throw "Failed to send data k_EResultLimitExceeded" errors in the console. It seems to only happen when a player is transferring into a separate zone via portal and the errors stop soon after. This led me down the rabbit hole looking for the cause of that error, and no one seems to have a definite solution for it either. I've noticed when a zone is hosted, the host can see an action executed by a peer in real-time, but not the peers. So the host may see a door open and close ten times before it's finally opened on the peer's side. I expect this is typical behavior, but I'm still unsure where the congestion is. Potentially an upload bandwidth bottleneck on the line? A host bandwidth bottleneck inside the game, necessitating the use of the client side DLL patch? Or even an issue with the massive number of instances being processed in heavily built or terraformed areas.

It's starting to feel a lot like playing desync whackamole, and each failed fix brings me closer to the conclusion that it's an issue on my end. It might be best in my case to move on to a VPS/hosting provider, or perhaps wait for the devs to sort things out.

Thank you for your contribution @rparker-indeed, I'll give it a shot as well.

**Reply by spannerman79** [reply]
You do know that the Valheim server isn't an actual dedicated sever - right?

Its a headless _client_.  So while there has been _limited_ improvements you are stil bound by P2P networking stack.  The one with the highest ping and the worst connection is the literal weakest link and brings everyone else down.

It would be like (for example) people in the West Cost of USA playing fine, all < 50ms pings, perfect internet then along comes me, from Australia, with a ping (at bare min to linode on the west cost I use to get between 214 - 240ms) of ~230ms. Which then causes everyone else to go from having nice <50ms pings to jump to my pings.

That is your wack-a-mole.  Find the weakest link (aka - the player with the most shitty connection)


Edit:  I'll even expand further on an incident that actually happened.

One user that connected in was using mobile based internet.  In Australia (and prob elsewhere too) mobile based internet access is **heavily** QoS and throttled.  This doesn't really affect http/https traffic but for time sen applications (VoIP, gaming, etc) this is a killer.  A _slight_ work around was to setup and use a ZeroTier network.  It was easy enough to setup, easy for others to install ZeroTier, add the network id, for me to log in and auth the client id etc.  Because the traffic of ZeroToier is _heavily_ encrypted (like a VPN) that user I mentioned beforehand wasn't causing everyone else to have weird issues to happen.  Their telco wasn't throttling or playing around with packet flow.

There is not as many issues as there once was (even _after_ the update that was pushed to improve dedicated server connections) - maybe 1 per hour of weird desync glitch now vs 1 per minute.

**Reply by TS-x** [reply]
Yes, I'm aware, but that's only half-true. There is no full P2P network stack, as the data is never goes between clients, only between the client/server. It's specifically an issue when the player **hosting** the zone has a bad connection, not when someone on the server or in the same zone has a bad connection. I've tried several different tests, including having my client hosting zones, which is on the same network as the server. The end result still the same. This is what originally led me to sendrate fixes.

Maybe the lines became blurred somewhere in my testing, but I'm almost certain this isn't exclusive to someone with packet loss or high ping just simply being connected.

**Reply by spannerman79** [reply]
With my testing (same game version, same V+ version), 5 users.  One (me) connecting in locally (LAN connection ~0.2ms), user 1 (Ok ISP, FTTH, 6ms normal, 30ms ping to me), connects (just them)- server pings are ~20ms.  User 2 (Budget ISP that uses CGNAT, FTTH, 6ms normal, 30 ms ping to me), connects(just them), server pings are ~50ms (I'd say CGNAT).  User 3 (Mobile network user, 12ms normal, 30ms ping to me), connects (just them), server pings are between 40ms up to 100ms

In this case User 4 is the weak link - so far.

User 1 and 2 (non mobile) connect in (without me) - pings are 20 - 30ms. 

So CGNAT connection with a non mobile carrier are ok.

Everyone connects in - server pings are literal nuts - jump between 30ms up to 150ms.

Two connections that use CGNAT do not play well with P2P.

So yes, P2P networking is still used, or some component of it.

Did the same test, same steps, this time connected in via ZeroTier.  Everyone connected in - pings ~20 - 30ms _including_ the mobile internet user


Edit:  I missed a user or two here and there - 4am here for me.  Basically - yes, ping **does** affect.  And the quality of the connection used to connect in.

Basically - CGNAT based connections are the worst and anyone that knows a bit about networking knows CGNAT and P2P do **not** play well together if CGNAT hasn't been configured correctly for IPv4 traffic and the ISP assigns the connecting user an IPv4 ip addy.
